benchmarks/bayesmark/report_bayesmark.py

```diff
@@ -0,0 +1,314 @@
+import abc
+from collections import defaultdict
+import io
+import itertools
+import os
+from typing import Dict
+from typing import Generator
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+
+import numpy as np
+import pandas as pd
+from scipy.special import binom
+from scipy.stats import mannwhitneyu
+
+
+_LINE_BREAK = "\n"
+_TABLE_HEADER = "|Ranking|Solver|"
+_HEADER_FORMAT = "|:---|---:|"
+_OVERALL_HEADER = "|Solver|Borda|Firsts|\n|:---|---:|---:|\n"
+
+Moments = Tuple[float, float]
+Samples = Dict[str, List[float]]
+
+
+class BaseMetric(object, metaclass=abc.ABCMeta):
+    @property
+    @abc.abstractmethod
+    def name(self) -> str:
+        """Metric name displayed in final report."""
+
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def calculate(self, data: pd.DataFrame) -> List[float]:
+        """Calculates metric for each study in data frame."""
+
+        raise NotImplementedError
+
+
+class BestValueMetric(BaseMetric):
+    name = "Best value"
+
+    def calculate(self, data: pd.DataFrame) -> List[float]:
+
+        return data.groupby("uuid").generalization.min().values
+
+
+class AUCMetric(BaseMetric):
+    name = "AUC"
+
+    def calculate(self, data: pd.DataFrame) -> List[float]:
+
+        aucs: List[float] = list()
+        for _, grp in data.groupby("uuid"):
+            auc = np.sum(grp.generalization.cummin() - grp.generalization.min())
```

Comment:
I'm not sure but it seems the value of `grp.generalization.min()` could differ between optimizations. If so, it would make it difficult to compare the meanings of multiple AUCs.
(e.g., if there is a sampler that always samples a constant parameter set, the value of `np.sum(grp.generalization.cummin() -  grp.generalization.min())` would always be 0 (the best AUC value))
 
How about using a common value (e.g, just a constant value `0`) here instead?
(please let me know if I misunderstand something 🙏 )

Reply:
> Having fixed reference would leak absolute generalization performance into metric

I think this is an　expected AUC characteristic.
Let me quote a paragraph about AUC from [the paper](https://proceedings.mlr.press/v64/dewancker_strategy_2016.pdf):
<img width="646" alt="image" src="https://user-images.githubusercontent.com/181413/172301702-345588d7-d738-437a-a134-7ebc4ba875cd.png">
It says "a specified lower bound f<sub>LB</sub> on the function" (i.e., a constant value for each function) is used to calculate AUC.

> AUC values could get very large since we are using MSE

Yeah, this is a subtle problem😅
This problem is due to an inherent property of MSE. So just accepting very large values in the result would be a ( compromised but) practical solution I think .

Reply:
Thank you for the explanation, makes sense :+1: . Implemented in e2d85c4f4de85d8cfff1f9db4b477fc729471189.